provider/pkg/provider/previewOutputs.go

```diff
@@ -0,0 +1,218 @@
+// and will be marked as computed
+func previewResourceOutputs(
+	inputs resource.PropertyMap,
+	types map[string]metadata.CloudAPIType,
+	outputs map[string]schema.PropertySpec,
+	readOnly []string,
+) resource.PropertyMap {
+	result := resource.PropertyMap{}
+	for name, prop := range outputs {
+		key := resource.PropertyKey(name)
+		if inputValue, ok := inputs[key]; ok {
+			result[key] = previewOutputValue(inputValue, types, &prop.TypeSpec, filterReadOnly(readOnly, name))
+			// if the property is a readOnly property, then it's an output only value
+			// and we should mark it as computed
+		} else if isReadOnly(readOnly, name) {
+			result[key] = resource.MakeComputed(resource.NewStringProperty(""))
+		}
+	}
+	return result
+}
+
+// isStableOutput determines if a property is a stable output or not.
+// This uses a very conservative heuristic and does not cover all cases.
+//
+// e.g. sometimes the arn property does not contain the full resource type name
+// - `aws-native:sagemaker:ModelExplainabilityJobDefinition` has a property `jobDefinitionArn`
+// - `aws-native:securityhub:PolicyAssociation` has a property `associationIdentifier`
+//   - TODO[pulumi/aws-native#1892]: in some cases this property is the `primaryIdentifier`. Could we use that as another heuristic?
+//     a readonly property that is also a primary identifier? It doesn't catch all cases, but would catch more
+func isStableOutput(propName string, resourceTypeName tokens.TypeName) bool {
+	typeName := resourceTypeName.String()
+	stableOutputsNameOnly := []string{
+		fmt.Sprintf("%sName", naming.ToSdkName(typeName)),
+		fmt.Sprintf("%sId", naming.ToSdkName(typeName)),
+		fmt.Sprintf("%sArn", naming.ToSdkName(typeName)),
+	}
+	// These are the most common properties that are stable outputs
+	// and are used to determine if an output is stable or not
+	topLevelStableOutputs := slices.Concat([]string{
+		"name",
+		"id",
+		"arn",
+	}, stableOutputsNameOnly)
+
+	// we can't handle arrays because we don't know which item in the array
+	// the value corresponds to
+	if isArrayProperty(propName) {
+		return false
+	}
+
+	// e.g. aws-native:s3:StorageLens has a storageLensConfiguration/StorageLensArn readOnly property
+	if strings.Contains(propName, "/") {
+		parts := strings.Split(propName, "/")
+		name := parts[len(parts)-1]
+		// If the property is a nested property then only consider it stable if
+		// the property contains the resource type name. There are a lot of cases where
+		// an object has a property called `id` or `arn` that is not  the resource id or arn
+		return slices.Contains(stableOutputsNameOnly, naming.ToSdkName(name))
+	}
+	return slices.Contains(topLevelStableOutputs, naming.ToSdkName(propName))
+}
+
+func isReadOnly(readOnly []string, key string) bool {
+	for _, prop := range readOnly {
+		if naming.ToSdkName(prop) == key {
+			return true
+		}
+	}
+	return false
+}
+
+func filterReadOnly(readOnly []string, key string) []string {
```

Comment:
This does a little bit more than filtering it seems.

Reply:
I'll add a comment